consensus/src/pipeline/pipeline_builder.rs

```diff
@@ -665,16 +710,28 @@ impl PipelineBuilder {
         commit_proof_fut: TaskFuture<LedgerInfoWithSignatures>,
         executor: Arc<dyn BlockExecutorTrait>,
         block: Arc<Block>,
-        enable_pre_commit: bool,
+        pre_commit_status: Arc<Mutex<PreCommitStatus>>,
     ) -> TaskResult<PreCommitResult> {
         let mut tracker = Tracker::start_waiting("pre_commit", &block);
         let (compute_result, _, _) = ledger_update_fut.await?;
         parent_block_pre_commit_fut.await?;
 
         order_proof_fut.await?;
 
-        if compute_result.has_reconfiguration() || !enable_pre_commit {
+        let wait_for_proof = {
+            let mut status_guard = pre_commit_status.lock();
+            let wait_for_proof = compute_result.has_reconfiguration() || !status_guard.is_active();
+            // it's a bit ugly here, but we want to make the check and update atomic in the pre_commit case
+            // to avoid race that check returns active, sync manager pauses pre_commit and round gets updated
+            if !wait_for_proof {
+                status_guard.update_round(block.round());
+            }
+            wait_for_proof
+        };
+
+        if wait_for_proof {
             commit_proof_fut.await?;
+            pre_commit_status.lock().update_round(block.round());
```

Comment:
can you do this?
``` rust
{
  let mut status_guard = pre_commit_status.lock();
  let wait_for_proof = compute_result.has_reconfiguration() || !status_guard.is_active();

  if wait_for_proof {
     drop(status_guard);
     commit_proof_fut.await?;
     // re-acquire lock
     pre_commit_status.lock().update_round(block.round());
  } else {
     status_guard.update_round(block.round());
  }
}
```

Reply:
that's exactly my code before, but compiler complains about holding the `status_guard` across the await point even I dropped it inside

Reply:
I have run into this issue before. The compiler doesn't think this is legal.